2021-11-monthly-report.org

```diff
@@ -0,0 +1,413 @@
+
+So overall removing the abstraction is going to make some things much easier to work with. For example, in the short term, code generation for unions/ADTs/match-expr could be simplified a lot if we remove this. It also might help attract more GCC people to work with the backend code generation piece to clean up the code here.
+
+See: https://github.com/Rust-GCC/gccrs/issues/412
+
+*** GCC TREE_ADDRESSABLE
+
+GCC requires VAR_DECL's and PARAM_DECL's to be marked with TREE_ADDRESSABLE when the declaration will be used in borrow's ('&' getting the address). This takes into account the implicit addresses when we do autoderef in method resolution/operator-overloading. TREE_ADDRESSABLE if not set allows the optimizers to use registers since no address in memory is required for this declaration, but this means we end up in cases like this:
+
+#+BEGIN_SRC rust
+#[lang = "add_assign"]
+pub trait AddAssign<Rhs = Self> {
+    fn add_assign(&mut self, rhs: Rhs);
+}
+
+impl AddAssign for i32 {
+    fn add_assign(&mut self, other: i32) {
+        *self += other
+    }
+}
+
+fn main() {
+    let mut a = 1;
+    a += 2;
+}
+#+END_SRC
+
+This generated GCC Generic IR such as:
+
+#+BEGIN_SRC rust
+i32 main ()
+{
+  i32 a.1; // <-- This is the copy
+  i32 D.86;
+  i32 a;
+
+  a = 1;
+  a.1 = a; // <-- Taking a copy
+
+  <i32 as AddAssign>::add_assign (&a.1, 2);
+  //                               ^
+  //                              ----
+
+  D.86 = 0;
+  return D.86;
+}
+#+END_SRC
+
+You can see GCC will automatically make a copy of the VAR_DECL resulting bad code-generation. But with the TREE_ADDRESSABLE set this looks like this:
+
+#+BEGIN_SRC rust
+i32 main ()
+{
+  i32 D.86;
+  i32 a;
+
+  a = 1;
+  <i32 as AddAssign>::add_assign (&a, 2);
+  D.86 = 0;
+  return D.86;
+}
+#+END_SRC
+
+The fix here now marks the declarations apropriately for when we need their address or not which then allows the GCC optimizers to work as we expect. For more info see this useful comment https://github.com/Rust-GCC/gccrs/blob/0024bc2f028369b871a65ceb11b2fddfb0f9c3aa/gcc/tree.h#L634-L649
+
+*** Qualified Path BugFix
+
+We found that the implementation of qualified paths in was reliant on some implictly injected names within the name-resolution process so that we could try and at least resolve the root of the qualified path. This implementation was never going to hold up but served as a simple hack to get the type system off the ground during the traits milestone. These hacks and implict names are now removed and qualified paths are now just like TypePaths resolved durin the type checking pass. The bug here was that the qualified path of "<Self as Foo>::A" was unable to resolve the root "<Self as Foo>" since no implicit name was generated here, but now the type system is able to properly project Self as Foo to then probe for A which means the type system is able to handle more complex qualified paths.
```

Comment:
```suggestion
We found that the implementation of qualified paths in was reliant on some implictly injected names within the name-resolution process so that we could try and at least resolve the root of the qualified path. This implementation was never going to hold up but served as a simple hack to get the type system off the ground during the traits milestone. These hacks and implicit names are now removed and qualified paths are now just like TypePaths resolved during the type checking pass. The bug here was that the qualified path of "<Self as Foo>::A" was unable to resolve the root "<Self as Foo>" since no implicit name was generated here, but now the type system is able to properly project Self as Foo to then probe for A which means the type system is able to handle more complex qualified paths.
```